R/bundlePackageRenv.R

```diff
@@ -0,0 +1,142 @@
+}
+
+parseRenvDependencies <- function(bundleDir, snapshot = FALSE) {
+  renv <- jsonlite::read_json(renvLockFile(bundleDir))
+
+  repos <- setNames(
+    vapply(renv$R$Repositories, "[[", "URL", FUN.VALUE = character(1)),
+    vapply(renv$R$Repositories, "[[", "Name", FUN.VALUE = character(1))
+  )
+  deps <- standardizeRenvPackages(
+    renv$Packages,
+    repos,
+    biocPackages = biocPackages(bundleDir)
+  )
+  if (nrow(deps) == 0) {
+    return(data.frame())
+  }
+
+  if (snapshot) {
+    # Can use system libraries
+    deps$description <- lapply(deps$Package, function(pkg) {
+      dcf <- read.dcf(system.file("DESCRIPTION", package = pkg))
+      as.list(as.data.frame(dcf))
+    })
+  } else {
+    old <- options(renv.verbose = FALSE)
+    defer(options(old))
+
+    # Generate a library from the lockfile
+    lib_dir <- dirCreate(file.path(bundleDir, "renv_library"))
+    renv::restore(bundleDir, library = lib_dir, prompt = FALSE)
+    defer(unlink(lib_dir, recursive = TRUE))
+
+    deps$description <- lapply(deps$Package, function(pkg) {
+      dcf <- read.dcf(file.path(lib_dir, pkg, "DESCRIPTION"))
+      as.list(as.data.frame(dcf))
+    })
+  }
+
+  deps
+}
+
+standardizeRenvPackages <- function(packages, repos, biocPackages = NULL) {
+  repos <- standardizeRepos(repos)
+  availablePackages <- availablePackages(repos)
+
+  names(packages) <- NULL
+  out <- lapply(
+    packages,
+    standardizeRenvPackage,
+    availablePackages = availablePackages,
+    biocPackages = biocPackages,
+    repos = repos
+  )
+  out <- compact(out)
+  out <- lapply(out, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(out)
+}
+
+standardizeRenvPackage <- function(pkg,
+                                   availablePackages,
+                                   biocPackages = NULL,
+                                   repos = character(),
+                                   bioc) {
+  # Don't include renv itself
+  if (identical(pkg$Package, "renv")) {
+    return(NULL)
+  }
+
+  # Convert renv source to manifest source/repository
+  # https://github.com/rstudio/renv/blob/0.17.2/R/snapshot.R#L730-L773
+
+  if (is.null(pkg$Repository) && !is.null(pkg$RemoteRepos) && grepl("bioconductor.org", pkg$RemoteRepos)) {
+    # Work around bug where renv fails to detect BioC package installed by pak
+    # https://github.com/rstudio/renv/issues/1202
+    pkg$Source <- "Bioconductor"
+  }
+
+  if (pkg$Source == "Repository") {
+    if (pkg$Repository == "CRAN") {
+      if (isDevVersion(pkg, availablePackages)) {
+        pkg$Source <- NA_character_
+        pkg$Repository <- NA_character_
+      } else {
+        pkg$Source <- "CRAN"
+        pkg$Repository <- findRepoUrl(pkg$Package, availablePackages)
+      }
+    } else {
+      # $Repository comes from DESCRIPTION and is set by repo, so can be
+      # anything. So we must look up from the package name
+      pkg$Repository <- findRepoUrl(pkg$Package, availablePackages)
```

Comment:
nit: during restore, renv first favors a matching repository name before falling back to "any" repository. not sure if we want that same pattern here or want to rely `available.packages` order like we do for Packrat.

Maybe this is: prefer the Repository named in renv.lock and prefer available.packages for snapshots.

Also worth considering: Should we prefer the renv.lock repository list or the list in the current R session?

https://github.com/rstudio/renv/blob/81fecb07fcc78cabb61b5309875e59484727b090/R/retrieve.R#L875-L880